rpc/gnmi/gnmi-config-subscriptions.md

```diff
@@ -0,0 +1,187 @@
+target to indicate a commit boundary to the client.
+
+## 2.1 Proto
+
+The extension contains a message called `ConfigSubscription` that carries
+one of 2 types of actions. `ConfigSubscriptionStart` or
+`ConfigSubscriptionSyncDone`
+
+```proto
+// ConfigSubscription extension allows clients to subscribe to configuration
+// schema nodes only.
+message ConfigSubscription {
+  oneof action {
+    // ConfigSubscriptionStart is sent by the client in the SubscribeRequest
+    ConfigSubscriptionStart start = 1;
+    // ConfigSubscriptionSyncDone is sent by the server in the SubscribeResponse
+    ConfigSubscriptionSyncDone sync_done = 2;
+  }
+}
+
+// ConfigSubscriptionStart is used to indicate to a target that for a given set
+// of paths in the SubscribeRequest, the client wishes to receive updates
+// for the configuration schema nodes only.
+message ConfigSubscriptionStart {}
+
+// ConfigSubscriptionSyncDone is sent by the server in the SubscribeResponse
+// after all the updates for the configuration schema nodes have been sent.
+message ConfigSubscriptionSyncDone {
+  // ID of a commit confirm operation as assigned by the client
+  // see Commit Confirm extension for more details.
+  string commit_confirm_id = 1;
+  // ID of a commit as might be assigned by the server
+  // when registering a commit operation.
+  string server_commit_id = 2;
+}
+```
+
+## 2.2 Actions
+
+### 2.2.1 ConfigSubscriptionStart
+
+A `ConfigSubscriptionStart` message is used by a gNMI client in a
+`SubscribeRequest` to indicate that it wants to start a ConfigSubscription.
+The target must respond exclusively with configuration data relevant to the
+created subscription.
+The target must respond with an initial set of updates followed by a
+`SubscribeResponse` where the `sync_update` field is set to `true`.
+However, if the `updates_only` field in the `SubscribeRequest` was set to
+`true`, the target should skip sending the initial updates and only send
+a `SubscribeResponse` with `sync_update` set to `true` as per the base gNMI
+specification.
+
+### 2.2.2 ConfigSubscriptionSyncDone
+
+A `ConfigSubscriptionSyncDone` message is used by a gNMI target in a
+`SubscribeResponse` to indicate a commit boundary to the client.
+A commit boundary marks the completion point of a specific set of
+configuration changes.
+It indicates that all changes within that set have been committed and
+that all notifications triggered by the commit have been streamed to
+the client.
+
+The `ConfigSubscriptionSyncDone` message includes two optional fields:
+
+* `commit_confirm_id`: A commit confirm ID assigned by the client which
+initiated the commit.
+The commit can be initiated via gNMI (using the CommitConfirmed Extension),
+NETCONF, or any other management interface. Applicable only if the commit
+confirmed option is used.
+* `server_commit_id`: An optional internal ID assigned by the target.
+
+In the case a commit happens before the `sync_response: true` and the server
```

Comment:
Adding a second thread here -- what happens if we have two configuration changes -- i.e.:

* client A makes a change
* client B makes a change
* client C has a `ConfigSubscription` open

It seems like, in this case, `sync_response: true` might be returned at any point in the lifetime of the subscription (when whatever existed in the data tree at the time client C created its subscription). We want `ConfigSubscriptionSyncDone` to be returned at the point that client A and client B have had their changes "committed" (let's clarify semantics here) -- this is rather independent to when `sync_response: true` is sent.

Another implication here is that we cannot tell the exact provenance of an update within the `Subscribe` stream if there are changes being made that do not block on the "completion" of a previous change.

Reply:
the config store does not need to support parallel commits for this situation to happen.

assume we take N seconds to stream all updates corresponding to a config change.

t=0 - config change 1 happens and is finished
t=1 - streaming for updates from config change 1 starts
t=Y (Y<N) - config change 2 happens and is finished
t=Y+1 - streaming for updates from config change 2 starts
t=N - streaming for updates from config change 1 ends
t=Y+N - streaming for updates from config change 2 ends

in this scenario -- it's the streaming of the updates that overlaps, not the commits. 

when the client receives the `ConfigSubscriptionSyncDone` with the commit ID of config change 1 in it, it may already have received updates for paths that were impacted only by config change 2 -- so the semantic that this marker means "you have the state that reflects the state only after change 1" is broken.

the only way to avoid this is by saying that the config store locks until all updates are streamed (i.e., N seconds) (which seems bad for performance), or there is a strict send buffer with no coalescing possible (which I don't think is the right behaviour -- and is already broken by existing gnmi implementations).

why does coalescing matter? well, if both change 1 and change 2 change `/foo/bar` and the update for `/foo/bar` hasn't been sent, then it's acceptable for an gNMI server to coalesce the two, and only send the value for `/foo/bar` that corresponds to the value after the second change.

i think the semantics that you're trying to define here for `ConfigSubscriptionSyncDone` are problematic with any kind of concurrency. to me, it's better to say that the gNMI server gives looser semantics about what it can guarantee -- i.e., it guarantees only that it is "done" with processing the changes that are covered by a particular commit. if the client *really* wants to ensure that its config subscription reflects exactly that intended state, it can implement locking between changes without adding complexity at the server side.

Reply:
I think this definition works: `it guarantees only that it is "done" with processing the changes that are covered by a particular commit.`. The streaming of changes depends on whether they were coalesced.

Reply:
ACK -- suggest to update the text here to describe that this is the semantic meaning of receiving a "done". I might also suggest adding the example of concurrency that means that this design decision is required in an appendix for future readers.